docs/design/2290-cadence-ndc.md

```diff
@@ -0,0 +1,310 @@
+
+In Cadence NDC, workflow history events are modeled as a tree, as shown in second example in [### Version History].
+
+Workflows which encounters divergence will have more than one history branches. Among all history branches, the history branch with the highest version is considered as the `current branch` and workflow mutable state is a summary of the current branch. Whenever there is a switch between workflow history branches, a complete rebuild of workflow mutable state will occur.
+
+
+### Zombie Workflows
+
+There is an existing contract that for any domain & workflow ID combination, there can be at most one run (domain & workflow ID & run ID) open / running.
+
+Cadence NDC aims to keep the workflow state as up-to-date as possible among all participating data centers.
+
+Due to the nature of Cadence NDC, i.e. workflow history events are replicated asynchronously, different run (same domain & workflow ID) can arrive at target data center at different time, sometimes out of order, as shown below:
+
+```
+| ------------- |          | ------------- |          | ------------- |
+| Data Center A |          | Network Layer |          | Data Center B |
+| ------------- |          | ------------- |          | ------------- |
+        |                          |                          |
+        | Run 1 Replication Events |                          |
+        | -----------------------> |                          |
+        |                          |                          |
+        | Run 2 Replication Events |                          |
+        | -----------------------> |                          |
+        |                          |                          |
+        |                          |                          |
+        |                          |                          |
+        |                          | Run 2 Replication Events |
+        |                          | -----------------------> |
+        |                          |                          |
+        |                          | Run 1 Replication Events |
+        |                          | -----------------------> |
+        |                          |                          |
+| ------------- |          | ------------- |          | ------------- |
+| Data Center A |          | Network Layer |          | Data Center B |
+| ------------- |          | ------------- |          | ------------- |
+```
+
+Since run 2 appears in data center B first, run 1 cannot be replicated as runnable due to rule `at most one run open` (see above), thus, `zombie` workflow state is introduced. Zombie state indicates a workflow which cannot be actively mutated by a data center (assuming corresponding domain is active in this data center). A zombie workflow can only be changed by replication task.
+
+Run 1 will be replicated similar as run 2, except run 1's workflow state will be zombie before run 1 reaches completion.
```

Comment:
Just curious, where do we keep this state? In DB?

Reply:
all persisted to database.

a workflow can potentially turn from a zombie workflow to current running workflow (in this data center), if say there is a data center which only has run 1 (no run 2) & the run 1 is running with a higher version.